### PR TITLE
[CALCITE-6620] VALUES created by RelBuilder do not have a homogeneous type

### PR DESCRIPTION
**Pull request by mihaibudiu**
The issue fixed here was first reported as part of https://issues.apache.org/jira/browse/CALCITE-6617

### CODE REVIEW

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java [hunk 1]
```diff
@@ -10754,6 +10757,17 @@ private void testCurrentDateFunc(Pair<String, Hook.Closeable> pair) {
     f0.forEachLibrary(libraries, consumer);
   }
 
+  @Test void differentTypeValues() {
```
Comment (julianhyde) [inline testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10760]: Does this test belong here? It isn't testing an operator, or use the fixture.

Could it be a Quidem test?

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 2]
```diff
@@ -2886,6 +2886,36 @@ private static boolean isGroupId(AggCall c) {
...
+   * respectively match the fields types of the row.
+   *
+   * @param cluster  Cluster
+   * @param rowType  Type expected for values
+   * @param values   A list of literals that should match the rowType */
+  private static List<RexLiteral> convertLiteralTypes(RelOptCluster cluster,
+      RelDataType rowType, List<RexLiteral> values) {
+    assert values.size() == rowType.getFieldCount();
+    RexExecutor executor = cluster.getPlanner().getExecutor();
+    if (executor == null) {
```
Comment (NobiGo) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2899]: Do we have a test for the `executor == null`？I'm not sure why this happened.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2899]: Who knows who is calling the builder and with what setup?
This is just to be safe.

File: testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java (hunk 1 above)
Comment (mihaibudiu) [reply testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java:10760]: I will move it.
I don't know if I can create the same environment in Quidem. A simple attempt to reproduce this bug with standard settings failed.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 2 above)
Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2899]: +1, getExecutor is a nullable method.

**Comment by mihaibudiu** [comment]
Several quidem tests now have extra spaces in strings because they use a type system without raggedUnionTypesToVarying. The extra spaces are introduced when unifying the types of the VALUES. The new results seem the right ones.

**Review by suibianwanwank — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 2 above)
Comment (NobiGo) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2899]: Maybe we can improve it by `RexExecutor rexImpl =
        Util.first(cluster.getPlanner().getExecutor(), RexUtil.EXECUTOR);`

File: core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml [hunk 3]
```diff
@@ -9174,6 +9174,17 @@ LogicalProject(DEPTNO=[$0], EXPR$1=[$1])
   LogicalSort(sort0=[$2], dir0=[ASC-nulls-first])
     LogicalProject(DEPTNO=[$7], EXPR$1=[RANK() OVER (PARTITION BY $0 ORDER BY $6 DESC NULLS LAST)], EXPR$2=[ROW_NUMBER() OVER (PARTITION BY $0 ORDER BY $6 NULLS FIRST)])
       LogicalTableScan(table=[[CATALOG, SALES, EMP]])
+]]>
+    </Resource>
```
Comment (rubenada) [inline core/src/test/resources/org/apache/calcite/test/SqlToRelConverterTest.xml:9178]: Do we need to add this to `SqlToRelConverterTest.xml` if we re not adding anything to `SqlToRelConverterTest.java` ?

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 4]
```diff
@@ -2896,6 +2896,9 @@ private static List<RexLiteral> convertLiteralTypes(RelOptCluster cluster,
       RelDataType rowType, List<RexLiteral> values) {
     assert values.size() == rowType.getFieldCount();
     RexExecutor executor = cluster.getPlanner().getExecutor();
+    if (executor == null) {
+      executor = RexUtil.EXECUTOR;
+    }
     if (executor == null) {
```
Comment (NobiGo) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2902]: Need to delete.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 5]
```diff
@@ -2886,6 +2886,39 @@ private static boolean isGroupId(AggCall c) {
...
+      executor = RexUtil.EXECUTOR;
+    }
+    if (executor == null) {
+      return values;
+    }
+
+    List<RelDataTypeField> fields = rowType.getFieldList();
+    List<RexLiteral> constants = new ArrayList<>();
+    RexBuilder builder = cluster.getRexBuilder();
+    RexSimplify simplify = new RexSimplify(builder, RelOptPredicateList.EMPTY, executor);
```
Comment (rubenada) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2909]: Nit: Could the RexSimplify be passed as a parameter of this method? That would make the code simpler & cleaner, and we would avoid (re)creating a RexSimplify on each iteration that calls `convertLiteralTypes`.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java (hunk 4 above)
Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2902]: I don't understand this comment.

Comment (rubenada) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2902]: I think @NobiGo means that the currently the patch code is:
```
if (executor == null) {
  executor = RexUtil.EXECUTOR;
}
if (executor == null) {
  return values;
}
```
The second `if` could be removed.

Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2902]: I'm not sure RexSimplify#simlify is necessarily needed here. if not, the user should be allowed to set Executor to NULL. Instead, I think
`RexExecutor rexImpl = Util.first(cluster.getPlanner().getExecutor(), RexUtil.EXECUTOR);`
is good

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2902]: If you don't simplify the expression won't be a RexLiteral

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:2902]: But turns out that the RelBuilder already contains a simplifier

**Review by rubenada — APPROVED** [review]

**Review by NobiGo — APPROVED** [review]

**Pull request merged**